File: php-symfony/src/Repository/LinkRepository.php
<?php

namespace App\Repository;

use App\Entity\Link;
use Doctrine\Bundle\DoctrineBundle\Repository\ServiceEntityRepository;
use Doctrine\Persistence\ManagerRegistry;

/**
 * Repository for the Link entity.
 *
 * Inherited methods from ServiceEntityRepository cover the common cases
 * (find, findOneBy, findBy, findAll). Custom query methods are added below
 * for operations that need JOINs or non-trivial DQL.
 *
 * @extends ServiceEntityRepository<Link>
 */
class LinkRepository extends ServiceEntityRepository
{
    public function __construct(ManagerRegistry $registry)
    {
        parent::__construct($registry, Link::class);
    }

    /**
     * Finds a link by its short code and eagerly loads all associated clicks
     * in a single LEFT JOIN query.
     *
     * Using addSelect('c') instead of a lazy load avoids the N+1 problem on
     * the stats endpoint, where every click would otherwise trigger a separate
     * SELECT.
     *
     * @param string $code The short code to look up
     *
     * @return Link|null The link with its clicks collection fully hydrated,
     *                   or null if no link with that code exists
     */
    public function findByCodeWithClicks(string $code): ?Link
    {
        return $this->createQueryBuilder('l')
            ->leftJoin('l.clicks', 'c')
            ->addSelect('c')
            ->where('l.code = :code')
            ->setParameter('code', $code)
            ->getQuery()
            ->getOneOrNullResult();
    }
}

File: java-springboot/src/main/java/com/chopchop/repository/LinkRepository.java
package com.chopchop.repository;

import com.chopchop.model.Link;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.Optional;

public interface LinkRepository extends JpaRepository<Link, Long> {
    Optional<Link> findByCode(String code);
    boolean existsByCode(String code);

    @Query("SELECT l FROM Link l LEFT JOIN FETCH l.clicks c WHERE l.code = :code ORDER BY c.clickedAt DESC")
    Optional<Link> findByCodeWithClicks(@Param("code") String code);
}
